db/db.go

```diff
@@ -4,40 +4,75 @@ import (
-var ErrNotFound = errors.New("key not found")
+// ErrKeyNotFound is returned when there is no value associated with a key.
+var ErrKeyNotFound = errors.New("key not found")
 
 // ErrEmptyKey is returned when key is empty.
 var ErrEmptyKey = errors.New("key cannot be empty")
 
 // ErrIndexOutOfRange is returned when the iterator index is not in a valid range.
 var ErrIndexOutOfRange = errors.New("iterator index out of range")
 
-// DB for storing key-value tuples. The key must be a string and the value must
-// be a byte slice.
-type DB interface {
-	// Insert a value associated with a key. This will override any existing
-	// value associated with the key.
-	Insert(key string, value []byte) error
+// Codec can do encoding/decoding between arbitrary data object and bytes.
+type Codec interface {
 
-	// Get the value associated with the key. If no value is associated with the
-	// key, then an `ErrNotFound` error will be returned.
-	Get(key string) ([]byte, error)
+	// Encode the object into a slice of bytes.
+	Encode(obj interface{}) ([]byte, error)
 
-	// Delete the value associated with the key.
-	Delete(key string) error
+	// Decode the bytes to its original data object and assign it to the given
+	// variable. Value underlying `value` must be a pointer to the correct
+	// type for object.
+	Decode(data []byte, value interface{}) error
 }
 
-// Iterable is a DB that can iterate over its key-value tuples.
-type Iterable interface {
-	DB
+// Table is a sql-like table for storing key-value pairs. It requires the key
+// to be a non-empty string and the value has the type which can be marshaled
+// and unmarshaled by the used Codec.
+type Table interface {
+
+	// Insert writes the key-value into the table.
+	Insert(key string, value interface{}) error
+
+	// Get the value associated with the given key and assign it to the given
+	// variable. This function returns ErrKeyNotFound if the key cannot be found.
+	Get(key string, value interface{}) error
 
-	// Size returns the number of key-value tuples in the Iterable DB.
+	// Delete the value with the given key from the table. It is safe to use
+	// this function to delete a key which is not in the table.
+	Delete(key string) error
+
+	// Size returns the number of data entries in the table.
 	Size() (int, error)
 
-	// Iterator returns an Iterator which can be used to iterate through all
-	// key-value tuples in the IterableDB.
-	Iterator() Iterator
+	// Iterator returns an iterator that can iterate over table.
+	Iterator() (Iterator, error)
+}
+
+// DB is a collection of tables. It allows user to maintain multiple tables
+// with the same underlying database driver. It will automatically creat a new
+// table when first time writing to it.
+type DB interface {
+	Table(name string) Table
+
+	// Insert the key-value pair into the table with given name. It will
+	// return `ErrEmptyKey` if the key is empty.
+	Insert(name string, key string, value interface{}) error
```

Comment:
This applies to `Insert`, `Get`, and `Delete` and relates to a previous comment about how the `Table` interface should work.

Ideally, the `DB` doesn't actually care about names when insert, deleting, and getting. A `DB` is at a lower-level than a `Table` and only cares about key/value pairs. The `Table` is an abstraction over the `DB` that enforces a particular type of pattern in the key (i.e. always have `{name}_` as a prefix).

This also implies that `DBs` should create `Tables` (e.g. not `myTable := myDB.Table("name")`) but rather a `Table` should be independently constructed around a `DB` (e.g. `myTable := table.New("name", myDB)`). This means that `DB` implementations do not need to know _anything_ about tables (specifically, will not need to store an in-mem map of them).